http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/conn/SdkTlsSocketFactory.java

```diff
@@ -82,7 +82,7 @@ protected final void prepareSocket(final SSLSocket socket) {
             }
         }
         if (target.size() > 0) {
-            String[] enabling = target.toArray(new String[target.size()]);
+            String[] enabling = target.toArray(new String[0]);
```

Comment:
Isn't target.size() a tiny bit more efficient?

Reply:
One would think! I was very suspicious of this recommendation by IntelliJ. However, this [extremely well-written, detailed, and thorough article](https://shipilev.net/blog/2016/arrays-wisdom-ancients/#_conclusion) explains why that is not the case, and hasn't been since mid-Java 6.

From the conclusion: `target.size()` cannot be predictively optimized the way the constant `0` can. And existing arrays need to be explicitly zeroed, but newly-allocated arrays can bypass the zeroing if the JVM knows it's going to be fully initialized.

Admittedly, this is a HotSpot-specific optimization, but I would hazard a guess that 75%+ of the SDK users are on HotSpot or a derivative (that includes Corretto, no?). This is also a very micro-optimization for this use case; I can't imagine there are a lot of enabled TLS protocols, or that the array is frequently copied.

Reply:
TIL. I didn't know it could avoid zeroing the array. Cool!